index.js

```
@@ -14,47 +13,90 @@ function createResolver(config, options) {
     resolve: resolve,
   };
 
-  var lastKey;
+  // Keep constants separately
+  var constants = {};
+
+  // Keep requested keys to detect (and disallow) recursive resolution
+  var stack = [];
 
   function resolve(key) {
-    if (typeof key !== 'string') {
-      return;
-    }
 
-    var appliedArgs = slice.call(arguments, 1);
+    if (constants.hasOwnProperty(key)) {
+      return constants[key];
+    }
 
     var definition = config[key];
     // Ignore options that are not defined
     if (!definition) {
       return;
     }
 
-    if (key === lastKey) {
+    if (stack.some(function(s) {
+      return s === key;
+    })) {
       throw new Error('Recursive resolution denied.');
     }
-    lastKey = key;
 
-    var option = options[key];
-    // Bind the option so it can resolve other options if necessary
-    if (typeof option === 'function') {
-      option = option.bind(resolver);
+    stack.push(key);
+    try {
+      var option = options[key];
+      var appliedArgs = slice.call(arguments, 1);
+      var args = [definition.type, option].concat(appliedArgs);
+
+      if (typeof option === 'function') {
+        option = normalize.apply(resolver, args);
+      } else {
+        option = undefined;
+      }
+
+      if (option === undefined) {
+        option = definition.default;
+        if (typeof option === 'function') {
+          option = option.apply(resolver, appliedArgs);
+        }
+      }
+
+      stack.pop();
+      return option;
+
+    } catch (err) {
```

If we **must** keep the `try`, this should probably be a finally, no?

---

I figured that we would not want to swallow any errors here, so I rethrow after cleaning up the stack.

---

Shouldn't post before first coffee, sorry. Yes `finally` would work and won't swallow any errors. Will amend.